python/shotgun_data/background_task_manager.py

```diff
@@ -0,0 +1,832 @@
+        loop for the thread but due to a but in Qt pre-4.8 we have to do some extra stuff to make
+        sure everything gets cleaned up properly!
+        """
+        # run the event loop:
+        self.exec_()
+
+        # before we quit, we need to move the worker back to the main thread.  This is to work around
+        # issues with Qt pre-4.8 where any QObject.deleteLater's aren't executed on thread exit
+        # which would result in the worker not being cleaned up correctly!
+        self._worker.moveToThread(QtCore.QCoreApplication.instance().thread())
+
+class BackgroundTaskManager(QtCore.QObject):
+    """
+    Main task manager class.  Manages a queue of tasks running them asyncronously through
+    a pool of worker threads.
+    """
+    # timeout when Shotgun connection fails
+    _SG_CONNECTION_TIMEOUT_SECS = 20
+
+    # signal emitted when a task has been completed
+    task_completed = QtCore.Signal(int, object, object)# uid, group, result
+    # signal emitted when a task fails for some reason
+    task_failed = QtCore.Signal(int, object, str, str)# uid, group, msg, traceback
+    # signal emitted when all tasks in a group have finished
+    task_group_finished = QtCore.Signal(object)# group
+
+    def __init__(self, parent, start_processing=False, max_threads=8):
+        """
+        Construction
+
+        :param parent:              The parent QObject for this instance
+        :param start_processing:    If True then processing of tasks will start immediately
+        :param max_threads:         The maximum number of threads the task manager will use at any
+                                    time.
+        """
+        QtCore.QObject.__init__(self, parent)
+
+        self._bundle = sgtk.platform.current_bundle()
+
+        self._next_task_id = 0
+        self._next_group_id = 0
+
+        self._can_process_tasks = start_processing
+
+        # the pending tasks, organized by priority
+        self._pending_tasks_by_priority = {}
+
+        # available threads and running tasks:
+        self._max_threads = max_threads or 8
+        self._all_threads = []
+        self._available_threads = []
+        self._running_tasks = {}
+
+        # various task look-up maps:
+        self._tasks_by_id = {}
+        self._group_task_map = {}
+
+        # track downstream dependencies for tasks:
+        self._upstream_task_map = {}
+        self._downstream_task_map = {}
+
+        self._threadlocal_storage = threading.local()
+
+    @property
+    def shotgun_connection(self):
+        """
+        Get a Shotgun connection to use.  Creates a new Shotgun connection if the instance doesn't 
+        already have one.  The connection returned should only be used by the current thread.
+
+        :returns:   The Shotgun connection for this instance in the current thread
+        """
+        if not hasattr(self._threadlocal_storage, "shotgun"):
```

Comment:
I'm not actually a fan of having this separate instance here and would much rather it just use the thread-specific instance from the core api...  

That said, this is here to match the pattern from the previous implementation of the data retriever that had its own shotgun instance.  The reason it moved to here is specifically to avoid each data retriever having a shotgun instance per thread (so num shotgun connections would have equaled num data retrievers x num threads) - however, I think it probably makes more sense to drop this completely and just use the shotgun instance from core.